keps/0006-lifecycle.md

```diff
@@ -0,0 +1,207 @@
+
+1. Test Harness control of KUTTL start with default configurations.  This includes which control plane to use, the number of concurrent tests and the location of TestSuites.
+2. TestSuite control which includes; applying manifests, CRDs, and running commands.  This enables a "before" testsuite model, however it is limited to the full collection of TestSuites.
+
+#### Flat TestSuites
+
+Today, KUTTL treats all configured testsuites (listed with `testDirs`) as 1 suite of tests.  While it makes senses for all tests to run under the given the established control plane, it is limiting for a configured testsuite to not be able to express a set of commands or manifests which are specific to that set of tests.  It currently requires the test harness configuration change based on intimate knowledge of the tests.  It would be great to separate out these concerns, which will result in more portable tests.
+
+
+## Proposal
+
+### Proposal: Component Definition
+
+![TestSuite Architecture](images/kuttl-testsuite.jpg)
+
+* **TestSuite Collection** - This is a collection of TestSuites, which could be 1 or more.  This traditionaly has been called a TestSuite.  The TestSuite Collection is configured with the test harness configuration file, `kuttl-test.yaml` by default. 
+* **TestSuite** - Is a folder which contains a collection of Tests defined by sub-folders.
+* ** Test** - Is a folder which contains a number of files which define the steps necessary to assert the correct of a test.  If a test folder has sub-folders they are not analyzed by KUTTL unless explicitly referenced through configuration or a teststep.
+* **TestStep** - Is the smallest component of a test, which are a collection of indexed files which govern they order of evaluation.  There are 4 basic types of TestStep files; 1) an apply manifest file (used for test setup), 2) an assert file, 3) an error file (used to assert the absense of a resource) and 4) a TestStep file which can run commands, update timeouts and delete resources. 
+
+### Proposal: Component Lifecycle
+
+When a test harness is run (be default with no flags for single test runs), the test harness will analyze all configured TestSuites establishing a memory model for execution (without analyzing every test / testfile).  A change from the current KUTTL, KUTTL will run each TestSuite as a unit (in order to support additional lifecycle hooks detailed in the next section). The order of TestSuite execution will be determined by the order of configuation in the kuttl-test.yaml configuration file.  Each TestSuite will run the set of tests defined by that suite with no guarantees on order.  Each Test will have a series of test steps which will run in index order.
+
+### Proposal: Component Lifecycle hooks
+
+It is proposed that the following activities are useful before and after certain test phases:
+
+1. apply manifests
+2. CRD management (create/delete)
+3. delete resources
+4. run commands
```

Comment:
Let's be clear that these activities are run in a specific order. Also, isn't CRD management included in the "apply manifests" and "delete resources" steps? Why does it have to be a separate step?

Reply:
CRDs are treated different in kuttl only in that there is a wait for the crd to be realized.... this is not true for manifests... any "wait" for arbitrary manifests is up to the user for an assert.